### DISCUSSION TITLE
use json array in liquid

### DISCUSSION THREAD

**Discussion by guidoDimasi** [elsa-workflows/elsa-core#5427] [Q&A]
i've a string like this "[ {"COD": "001", "DESC": "Direttore commerciale","EMAIL": "prova@dimasi.it"}, { "COD": "01", "DESC": "Cesare Dimasi", "EMAIL": "mail@ordini.it" } ]"

is it possible to convert into a Json array? so it could be used i loops or email template?

keep in mind that i'm using Workflow Designer

**Comment by glime-ai** [comment]
Yes, the string you've provided is already in JSON array format, which is a common way to represent a list of objects in JSON (JavaScript Object Notation). You can use this JSON array in loops or email templates within a workflow designed with Elsa Workflow Designer.

To use this JSON array in a loop, you would typically use a "ForEach" activity to iterate over each item in the array. Within the loop, you can access the properties of the current item, such as "COD", "DESC", and "EMAIL".

For email templates, you can use the values from the JSON array to dynamically populate fields in the email, such as the recipient's email address or the email body content.

Here's a general outline of how you might use this JSON array in Elsa Workflow Designer:

1. Define the JSON array as a workflow variable or obtain it from an activity's output.
2. Add a "ForEach" activity to your workflow and configure it to iterate over the JSON array.
3. Inside the "ForEach" loop, add activities that utilize the current item's properties. For example, you could use a "SendEmail" activity and set the "To" property to the current item's "EMAIL" field.
4. If you're using the JSON array in an email template, you can reference the properties using the appropriate syntax for the template language you're using (e.g., Liquid syntax).

Remember to ensure that the JSON is valid and properly formatted for use in the workflow. If you encounter any issues or errors while setting this up in Elsa Workflow Designer, you can refer to the Elsa documentation or seek help from the community forums or support channels.

Would you like a more detailed example of how to set this up in Elsa Workflow Designer, or do you need assistance with a specific part of the process?

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).

**Comment by brodrigz** [comment]
I've also ran into this issue and had to resort to creating custom activities, which is a shame.

The Liquid syntax allows exactly for this type of use case, but its current implementation into elsa doesn't allow for seamless iteration on json arrays.

I can have an Object[] or JSON[], or JsonElement[] variable lets say like this:
```
[
{"PropX":"1",
"PropY":"2"},
{"PropX":"3",
"PropY":"4"}
]

```

In a liquid expression I CAN iterate over them:
```
{% for item in Variables.my_variable%}
   Item: {{item}}
{% endfor %}
```

This would get compiled into:
```
Item: {"PropX":"1","PropY":"2"}
Item: {"PropX":"3","PropY":"4"}
```

However we cannot access any of the object's properties
{{item.PropX}} doesn't work
{{item['PropX']} doesn't work also

If it did work it would be perfect for easy email templating on a single field, without having to rely on the ForEach activity.

**Comment by sfmskywalker** [comment]
Yes — the reliable approach is to parse the JSON string into a JavaScript array/object first, then store that parsed value in a workflow variable. Liquid can iterate the array and access members when it receives that JSON-backed shape; it cannot convert a JSON *string* into objects by itself.

In Designer, for example, keep the text in a `JsonText` variable and populate an `Items` variable using a **Run JavaScript** activity:

```javascript
setItems(JSON.parse(getJsonText()))
```

Then a Liquid template can use the parsed variable:

```liquid
{% for item in Variables.Items %}
  {{ item.COD }} — {{ item.DESC }} — {{ item.EMAIL }}
{% endfor %}
```

The current integration tests cover this shape: a JavaScript activity stores an array of objects in a workflow variable, and Liquid reads values such as `Variables.Products[0].id` and `.price`. The Liquid integration explicitly registers member access for `ExpandoObject` and for declared variable types, so an arbitrary `object[]` or `JsonElement[]` can behave differently from the parsed workflow variable.

If that still does not expose `item.COD`, please share your Elsa version, the variable's declared type/storage driver, and a minimal workflow export. That will help distinguish a runtime-shape issue from a Liquid configuration issue.

**Comment by sfmskywalker** [comment]
Closing this as answered. If the parsed workflow variable still does not expose item members in Liquid, please open a new discussion with the Elsa version, variable type/storage driver, and a minimal workflow export.
